File: customer_app/lib/models/product.dart
class Product {
  final int id;
  final int sellerId;
  final int? categoryId;
  final bool isOther;
  final String name;
  final String? description;
  final double price;
  final String unit;
  final double availableQty;
  final double minOrderQty;
  final String stockStatus;
  final String? imageUrl;
  final bool isActive;
  final String? sellerShopName; // present only if backend embeds seller info

  Product({
    required this.id,
    required this.sellerId,
    this.categoryId,
    this.isOther = false,
    required this.name,
    this.description,
    required this.price,
    required this.unit,
    required this.availableQty,
    required this.minOrderQty,
    required this.stockStatus,
    this.imageUrl,
    this.isActive = true,
    this.sellerShopName,
  });

  bool get inStock => stockStatus.toLowerCase() != 'out_of_stock' && availableQty > 0;

factory Product.fromJson(Map<String, dynamic> json) {
    double num(dynamic v) => v == null ? 0 : (v is int || v is double ? v.toDouble() : double.tryParse(v.toString()) ?? 0);
    return Product(
      id: json['id'],
      sellerId: json['seller_id'] ?? 0,
      categoryId: json['category_id'],
      isOther: json['is_other'] ?? (json['category_id'] == null),
      name: json['name']?.toString() ?? '',
      description: json['description']?.toString(),
      price: num(json['price']),
      unit: json['unit']?.toString() ?? '',
      availableQty: num(json['available_qty']),
      minOrderQty: num(json['min_order_qty'] ?? 1),
      stockStatus: json['stock_status']?.toString() ?? 'in_stock',
      imageUrl: json['image_url']?.toString(),
      isActive: json['is_active'] ?? true,
      sellerShopName: json['seller']?['shop_name']?.toString() ?? json['shop_name']?.toString(),
    );
  }
}

File: backend/app/models/product.py
import enum
from sqlalchemy import Column, Integer, String, Float, Boolean, ForeignKey, Enum, Text
from sqlalchemy.orm import relationship

from app.core.database import Base


class StockStatus(str, enum.Enum):
    in_stock = "in_stock"
    low_stock = "low_stock"
    out_of_stock = "out_of_stock"


class Product(Base):
    __tablename__ = "products"

    id = Column(Integer, primary_key=True, index=True)
    seller_id = Column(Integer, ForeignKey("sellers.id"), nullable=False)

    # nullable category_id = product is currently "Other / Uncategorized"
    # (requirement #14) — admin later moves it into a real category via
    # /products/{id}/move-category, which just sets this FK — original
    # product row is never deleted (requirement #20).
    category_id = Column(Integer, ForeignKey("categories.id"), nullable=True, index=True)
    is_other = Column(Boolean, default=False)  # True until admin moves it to a category

    name = Column(String(150), nullable=False, index=True)
    description = Column(Text, nullable=True)
    price = Column(Float, nullable=False)
    unit = Column(String(30), nullable=False)  # piece/kg/gram/litre/packet/etc (free text, admin-configurable list on client)
    available_qty = Column(Float, default=0)
    min_order_qty = Column(Float, default=1)
    stock_status = Column(Enum(StockStatus), default=StockStatus.in_stock, index=True)
    image_url = Column(String(255), nullable=True)
    is_active = Column(Boolean, default=True)

    seller = relationship("Seller", back_populates="products")
    category = relationship("Category", back_populates="products")
    keywords = relationship("ProductKeyword", back_populates="product", cascade="all, delete-orphan")


class ProductKeyword(Base):
    """
    Extra search terms a seller enters (requirement #14/#15) — e.g. for
    चप्पल: 'फुटवियर', 'footwear', 'slipper', 'sandal'. Indexed so the
    smart search can match on any of them.
    """
    __tablename__ = "product_keywords"

    id = Column(Integer, primary_key=True, index=True)
    product_id = Column(Integer, ForeignKey("products.id"), nullable=False)
    keyword = Column(String(100), nullable=False, index=True)

    product = relationship("Product", back_populates="keywords")
